Separate elementwise level2 math functions

xiaomengy opened the pull request:
Summary:
Separate elementwise level2 math functions

i-am-not-moving-c2-to-c10

Differential Revision: D13954928

houseroad commented:
Ah... conflicted... you have to rebase again...

houseroad reviewed (approved):
Looks good. Thanks for splitting the math related files.

Do we plan to move the BroadcastBinaryOp (something like `math::Add(ndim, A_dims, B_dims, C_dims, op, A, B, C)`) to elementwise.h or broadcast.h? I feel broadcast.h is more appropriate.

BIT-silence commented on caffe2/utils/math/elementwise.cc:
```diff
@@ -202,74 +238,155 @@ DELEGATE_ERF_FUNCTION(double)
+    EigenVectorMap<T>(C, N) = ConstEigenVectorArrayMap<T>(A, N)               \
+        EigenOp ConstEigenVectorArrayMap<T>(B, N);                            \
+  }
+DELEGATE_SIMPLE_BINARY_FUNCTION_BY_EIGEN_OPERATOR(std::int32_t, Add, +)
```
line 212 is under #ifdef CAFFE2_MKL #else part, which means if there is no MKL, this part is global since no MKL support for int types.

houseroad replied:
I see, that makes sense.

xiaomengy committed:
Separate elementwise level2 math functions (#16753)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/16753

Separate elementwise level2 math functions

i-am-not-moving-c2-to-c10

Differential Revision: D13954928

fbshipit-source-id: ba43cc32fe65960cac067a0bcededffee66da199
